src/nlv_object.h

```diff
@@ -5,182 +5,184 @@
+    }
+    if(val->IsArray()) {
+      Local<Array> flags_ = Local<Array>::Cast(val);
+      unsigned int flags = 0;
+      for (unsigned int i = 0; i < flags_->Length(); i++)
+        flags |= flags_->Get(Nan::New<Integer>(i))->Int32Value();
+      return flags;
+    } else if(val->IsNumber()) {
+      return val->IntegerValue();
+    } else {
+      Nan::ThrowTypeError("flags must be an array or a number");
+      return 0;
     }
-
-    return ref_;
-  }
-
-  void SetInvalid() {
-    ref_ = NULL;
-    valid_ = false;
   }
 
-protected:
-  NLVObjectBase *ref_;
-  bool valid_;
-};
-
-#define NLV_STRINGIFY0(v) #v
-#define NLV_STRINGIFY(v) NLV_STRINGIFY0(v)
-
-// macro could be removed but alternate solution seems overly complex
-// https://blog.molecular-matters.com/2015/12/11/getting-the-type-of-a-template-argument-as-string-without-rtti/
-#define NLV_OBJECT_STATIC_HELPERS(class_name) \
-  static const char* ClassName() { \
-    return NLV_STRINGIFY(class_name); \
-  } \
-  friend class NLVObject;
+  template <typename ParentClass, typename HandleType, typename CleanupHandler>
+  class NLVObject : public NLVObjectBase
+  {
+  public:
+    typedef HandleType handle_type;
 
+    typedef typename std::remove_pointer<HandleType>::type HandleValue;
 
-class NLVObjectBase : public Nan::ObjectWrap
-{
-public:
-  virtual void ClearHandle() = 0;
-  virtual void ClearChildren() = 0;
-  virtual void SetParentReference(NLVObjectBasePtr *parentReference) = 0;
+    NLVObject(HandleType handle) : handle_(handle, CleanupHandler::cleanup), parentReference_(NULL) {}
+    ~NLVObject() {
+      // calling virtual ClearHandle() will break if overridden by subclasses
+      ClearHandle();
+      if (parentReference_ != NULL) {
+        parentReference_->SetInvalid();
+      }
+    }
 
-  std::vector<NLVObjectBasePtr*> children_;
-};
+    struct MethodNoReturnHandler {
+      template<typename T>
+      Worker::OnFinishedHandler operator()(T val) {
+        return NLV::PrimitiveReturnHandler(true);
+      }
+    };
 
-template <typename ParentClass, typename HandleType, typename CleanupHandler>
-class NLVObject : public NLVObjectBase
-{
-public:
-  typedef HandleType handle_type;
+    template<typename T>
+    struct MethodReturn {
+      Worker::OnFinishedHandler operator()(T val) {
+        return NLV::PrimitiveReturnHandler(static_cast<T>(val));
+      }
+    };
+    using MethodReturnBool = MethodReturn<bool>;
+    using MethodReturnInt = MethodReturn<int>;
+
+    struct MethodReturnString {
+      Worker::OnFinishedHandler operator()(const char* val) {
+        std::string str = val;
+        return [=](Worker* worker) {
+          Nan::HandleScope scope;
+          v8::Local<v8::Value> argv[] = { Nan::Null(), Nan::New(str).ToLocalChecked() };
+          worker->Call(2, argv);
+        };
+      }
+    };
+    
+    static inline bool virHasFailed(int val) {
+      return val < 0;
+    }
+    
+    static inline bool virHasFailed(const void* ptr) {
```

Comment:
Is it? It's a bit difficult to follow with the use of `auto` down there, but it looks like `virHasFailed` is being used to replace the old `if (result == -1)` calls everywhere.  So [this](https://github.com/hooklift/node-libvirt/pull/77/files#diff-c679822ceba701f71305a5025962b67aR103) I think is the result of the vir method being called, which tends to be an integer - maybe because you're using a `const void*` signature here you're unintentionally following this path?